// thirdweb/dist/cjs/wallets/wallet-connect/consumer/request-handlers/send-transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleSendTransactionRequest = void 0;
const hex_js_1 = require("../../../../utils/encoding/hex.js");
const utils_js_1 = require("../utils.js");
/**
 * @internal
 */
async function handleSendTransactionRequest(options) {
    const { account, chainId, params: [transaction], } = options;
    if (transaction.from !== undefined) {
        (0, utils_js_1.validateAccountAddress)(account, transaction.from);
    }
    const txResult = await account.sendTransaction({
        gas: transaction.gas ? (0, hex_js_1.hexToBigInt)(transaction.gas) : undefined,
        gasPrice: transaction.gasPrice
            ? (0, hex_js_1.hexToBigInt)(transaction.gasPrice)
            : undefined,
        value: transaction.value ? (0, hex_js_1.hexToBigInt)(transaction.value) : undefined,
        nonce: transaction.nonce ? (0, hex_js_1.hexToNumber)(transaction.nonce) : undefined,
        to: transaction.to,
        data: transaction.data,
        chainId,
    });
    return txResult.transactionHash;
}
exports.handleSendTransactionRequest = handleSendTransactionRequest;
