// thirdweb/dist/cjs/extensions/erc721/__generated__/IBatchMintMetadata/read/getBaseURICount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBaseURICount = exports.decodeGetBaseURICountResult = exports.isGetBaseURICountSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x63b45e2d";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getBaseURICount` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getBaseURICount` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isGetBaseURICountSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isGetBaseURICountSupported(contract);
 * ```
 */
async function isGetBaseURICountSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetBaseURICountSupported = isGetBaseURICountSupported;
/**
 * Decodes the result of the getBaseURICount function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC721
 * @example
 * ```ts
 * import { decodeGetBaseURICountResult } from "thirdweb/extensions/erc721";
 * const result = decodeGetBaseURICountResult("...");
 * ```
 */
function decodeGetBaseURICountResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetBaseURICountResult = decodeGetBaseURICountResult;
/**
 * Calls the "getBaseURICount" function on the contract.
 * @param options - The options for the getBaseURICount function.
 * @returns The parsed result of the function call.
 * @extension ERC721
 * @example
 * ```ts
 * import { getBaseURICount } from "thirdweb/extensions/erc721";
 *
 * const result = await getBaseURICount({
 *  contract,
 * });
 *
 * ```
 */
async function getBaseURICount(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getBaseURICount = getBaseURICount;
