// thirdweb/dist/cjs/extensions/erc721/__generated__/IDrop/read/baseURIIndices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseURIIndices = exports.decodeBaseURIIndicesResult = exports.encodeBaseURIIndices = exports.encodeBaseURIIndicesParams = exports.isBaseURIIndicesSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xd860483f";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "index",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `baseURIIndices` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `baseURIIndices` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isBaseURIIndicesSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isBaseURIIndicesSupported(contract);
 * ```
 */
async function isBaseURIIndicesSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isBaseURIIndicesSupported = isBaseURIIndicesSupported;
/**
 * Encodes the parameters for the "baseURIIndices" function.
 * @param options - The options for the baseURIIndices function.
 * @returns The encoded ABI parameters.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeBaseURIIndicesParams } "thirdweb/extensions/erc721";
 * const result = encodeBaseURIIndicesParams({
 *  index: ...,
 * });
 * ```
 */
function encodeBaseURIIndicesParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.index]);
}
exports.encodeBaseURIIndicesParams = encodeBaseURIIndicesParams;
/**
 * Encodes the "baseURIIndices" function into a Hex string with its parameters.
 * @param options - The options for the baseURIIndices function.
 * @returns The encoded hexadecimal string.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeBaseURIIndices } "thirdweb/extensions/erc721";
 * const result = encodeBaseURIIndices({
 *  index: ...,
 * });
 * ```
 */
function encodeBaseURIIndices(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeBaseURIIndicesParams(options).slice(2));
}
exports.encodeBaseURIIndices = encodeBaseURIIndices;
/**
 * Decodes the result of the baseURIIndices function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC721
 * @example
 * ```ts
 * import { decodeBaseURIIndicesResult } from "thirdweb/extensions/erc721";
 * const result = decodeBaseURIIndicesResult("...");
 * ```
 */
function decodeBaseURIIndicesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeBaseURIIndicesResult = decodeBaseURIIndicesResult;
/**
 * Calls the "baseURIIndices" function on the contract.
 * @param options - The options for the baseURIIndices function.
 * @returns The parsed result of the function call.
 * @extension ERC721
 * @example
 * ```ts
 * import { baseURIIndices } from "thirdweb/extensions/erc721";
 *
 * const result = await baseURIIndices({
 *  contract,
 *  index: ...,
 * });
 *
 * ```
 */
async function baseURIIndices(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.index],
    });
}
exports.baseURIIndices = baseURIIndices;
