// thirdweb/dist/cjs/wallets/coinbase/coinbaseMobileSDK.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initMobileProvider = void 0;
const wallet_mobile_sdk_1 = require("@coinbase/wallet-mobile-sdk");
async function initMobileProvider(args) {
    (0, wallet_mobile_sdk_1.configure)({
        callbackURL: new URL(args.callbackURL || "https://thirdweb.com/wsegue"),
        hostURL: new URL(args.hostURL || "https://wallet.coinbase.com/wsegue"),
        hostPackageName: args.hostPackageName || "org.toshi",
    });
    let CoinbaseWalletMobileSDK = (await Promise.resolve().then(() => require(
    // @ts-ignore this is only visible from RN
    "@coinbase/wallet-mobile-sdk/build/WalletMobileSDKEVMProvider"))).WalletMobileSDKEVMProvider;
    if (typeof CoinbaseWalletMobileSDK !== "function" &&
        // @ts-ignore this is only visible from RN
        typeof CoinbaseWalletMobileSDK.default === "function") {
        CoinbaseWalletMobileSDK = CoinbaseWalletMobileSDK.default;
    }
    return new CoinbaseWalletMobileSDK({
        jsonRpcUrl: args.chain?.rpc,
        chainId: args.chain?.id,
    });
}
exports.initMobileProvider = initMobileProvider;
