// thirdweb/dist/cjs/extensions/erc721/__generated__/IDelayedReveal/read/encryptedData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encryptedData = exports.decodeEncryptedDataResult = exports.encodeEncryptedData = exports.encodeEncryptedDataParams = exports.isEncryptedDataSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa05112fc";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "index",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes",
    },
];
/**
 * Checks if the `encryptedData` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `encryptedData` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isEncryptedDataSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isEncryptedDataSupported(contract);
 * ```
 */
async function isEncryptedDataSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isEncryptedDataSupported = isEncryptedDataSupported;
/**
 * Encodes the parameters for the "encryptedData" function.
 * @param options - The options for the encryptedData function.
 * @returns The encoded ABI parameters.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeEncryptedDataParams } "thirdweb/extensions/erc721";
 * const result = encodeEncryptedDataParams({
 *  index: ...,
 * });
 * ```
 */
function encodeEncryptedDataParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.index]);
}
exports.encodeEncryptedDataParams = encodeEncryptedDataParams;
/**
 * Encodes the "encryptedData" function into a Hex string with its parameters.
 * @param options - The options for the encryptedData function.
 * @returns The encoded hexadecimal string.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeEncryptedData } "thirdweb/extensions/erc721";
 * const result = encodeEncryptedData({
 *  index: ...,
 * });
 * ```
 */
function encodeEncryptedData(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeEncryptedDataParams(options).slice(2));
}
exports.encodeEncryptedData = encodeEncryptedData;
/**
 * Decodes the result of the encryptedData function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC721
 * @example
 * ```ts
 * import { decodeEncryptedDataResult } from "thirdweb/extensions/erc721";
 * const result = decodeEncryptedDataResult("...");
 * ```
 */
function decodeEncryptedDataResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeEncryptedDataResult = decodeEncryptedDataResult;
/**
 * Calls the "encryptedData" function on the contract.
 * @param options - The options for the encryptedData function.
 * @returns The parsed result of the function call.
 * @extension ERC721
 * @example
 * ```ts
 * import { encryptedData } from "thirdweb/extensions/erc721";
 *
 * const result = await encryptedData({
 *  contract,
 *  index: ...,
 * });
 *
 * ```
 */
async function encryptedData(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.index],
    });
}
exports.encryptedData = encryptedData;
