// thirdweb/dist/cjs/wallets/wallet-connect/consumer/session-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setWalletConnectSessions = exports.removeSession = exports.saveSession = exports.initializeSessionStore = exports.getSessions = exports.walletConnectSessions = void 0;
const LocalStorage_js_1 = require("../../in-app/web/utils/Storage/LocalStorage.js");
/**
 * @internal
 */
async function getSessions() {
    if (!exports.walletConnectSessions) {
        return [];
    }
    const stringifiedSessions = await exports.walletConnectSessions.getWalletConnectSessions();
    return JSON.parse(stringifiedSessions ?? "[]");
}
exports.getSessions = getSessions;
/**
 * @internal
 */
function initializeSessionStore(options) {
    if (!exports.walletConnectSessions) {
        exports.walletConnectSessions = new LocalStorage_js_1.LocalStorage({
            clientId: options.clientId,
        });
    }
}
exports.initializeSessionStore = initializeSessionStore;
/**
 * @internal
 */
async function saveSession(session) {
    if (!exports.walletConnectSessions) {
        return;
    }
    const stringifiedSessions = await exports.walletConnectSessions.getWalletConnectSessions();
    const sessions = JSON.parse(stringifiedSessions ?? "[]");
    sessions.push(session);
    await exports.walletConnectSessions.saveWalletConnectSessions(JSON.stringify(sessions));
}
exports.saveSession = saveSession;
/**
 * @internal
 */
async function removeSession(session) {
    if (!exports.walletConnectSessions) {
        return;
    }
    const stringifiedSessions = await exports.walletConnectSessions.getWalletConnectSessions();
    const sessions = JSON.parse(stringifiedSessions ?? "[]");
    const newSessions = sessions.filter((s) => s.topic !== session.topic);
    await exports.walletConnectSessions.saveWalletConnectSessions(JSON.stringify(newSessions));
}
exports.removeSession = removeSession;
/**
 * @internal FOR TESTING ONLY
 */
function setWalletConnectSessions(storage) {
    exports.walletConnectSessions = storage;
}
exports.setWalletConnectSessions = setWalletConnectSessions;
