// thirdweb/dist/cjs/wallets/wallet-connect/consumer/session-proposal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.acceptSessionProposal = exports.disconnectExistingSessions = exports.onSessionProposal = void 0;
const index_js_1 = require("./index.js");
const session_store_js_1 = require("./session-store.js");
/**
 * @internal
 */
async function onSessionProposal(options) {
    const { wallet, walletConnectClient, event, onConnect } = options;
    const account = wallet.getAccount();
    if (!account) {
        throw new Error("[WalletConnect] No account connected to provided wallet");
    }
    const origin = event.verifyContext?.verified?.origin;
    if (origin) {
        await disconnectExistingSessions({ origin, walletConnectClient });
    }
    const session = await acceptSessionProposal({
        account,
        walletConnectClient,
        sessionProposal: event,
    });
    await (0, session_store_js_1.saveSession)(session);
    wallet.subscribe("disconnect", () => {
        (0, index_js_1.disconnectWalletConnectSession)({ session, walletConnectClient });
    });
    onConnect?.(session);
}
exports.onSessionProposal = onSessionProposal;
/**
 * @internal
 */
async function disconnectExistingSessions({ walletConnectClient, origin, }) {
    const sessions = await (0, session_store_js_1.getSessions)();
    for (const session of sessions) {
        if (session.origin === origin) {
            await (0, index_js_1.disconnectWalletConnectSession)({ session, walletConnectClient });
        }
    }
}
exports.disconnectExistingSessions = disconnectExistingSessions;
/**
 * @internal
 */
async function acceptSessionProposal({ account, walletConnectClient, sessionProposal, }) {
    if (!sessionProposal.params.requiredNamespaces.eip155) {
        throw new Error("[WalletConnect] No EIP155 namespace found in Wallet Connect session proposal");
    }
    if (!sessionProposal.params.requiredNamespaces.eip155.chains) {
        throw new Error("[WalletConnect] No chains found in EIP155 Wallet Connect session proposal namespace");
    }
    const namespaces = {
        chains: [
            ...sessionProposal.params.requiredNamespaces.eip155.chains.map((chain) => `${chain}:${account.address}`),
            ...(sessionProposal.params.optionalNamespaces?.eip155?.chains?.map((chain) => `${chain}:${account.address}`) ?? []),
        ],
        methods: [
            ...sessionProposal.params.requiredNamespaces.eip155.methods,
            ...(sessionProposal.params.optionalNamespaces?.eip155?.methods ?? []),
        ],
        events: [
            ...sessionProposal.params.requiredNamespaces.eip155.events,
            ...(sessionProposal.params.optionalNamespaces?.eip155?.events ?? []),
        ],
    };
    const approval = await walletConnectClient.approve({
        id: sessionProposal.id,
        namespaces: {
            eip155: {
                accounts: namespaces.chains,
                methods: namespaces.methods,
                events: namespaces.events,
            },
        },
    });
    const session = await approval.acknowledged();
    return {
        topic: session.topic,
        origin: sessionProposal.verifyContext?.verified?.origin || "Unknown origin",
    };
}
exports.acceptSessionProposal = acceptSessionProposal;
